### ISSUE TITLE
Addition of classes to assist using SymEngine in numerical codes

### ISSUE THREAD

**Issue by jppelteret** [symengine/symengine#1502]
Hi all,

I'm one of the developers of the [deal.II finite element library](http://dealii.org/), and have been working towards integrating SymEngine support into our codebase. For this to be possible, we've needed to implement what amounts to a C++ wrapper for SymEngine numbers that supports "traditional" operator overloading as well as casting (i.e. self-evaluation). It also understands a certain subset of features that SymEngine provides that are useful within the context of finite element methods. Due to the manner in which we intend to employ SymEngine, @isuruf and I have developed a wrapper to the various optimizer classes as well, which is flexible (i.e. one can select at runtime which optimizer to use, whether or not to invoke CSE, etc.) and works well for the simultaneous evaluation of the many lengthy expressions that result from the computation of constitutive model linearisations, finite element linearisations, etc.

I'm sure that we are not the only numerical code that would benefit from these sort of data structures, so I would like to propose integrating these base classes into SymEngine itself. Specifically, I propose that we place the source code in a directory `symengine/numerics` and wrap the implementation in the namespace `SymEngine::Numerics`. 

So far, I have the following to contribute:
- [ ] `numerics/types.h`: Some (more expressive) type definitions that make their contextual use clear.
- [ ] `numerics/utilities.[h/cpp]`: Some useful functions, including helper functions / structs that assist in making a unified interface to SymEngine number types possible.
- [ ] `numerics/number_types.[h/cpp]`: A wrapper to SymEngine numbers and functions used in the context of numerical codes.
- [ ] `numerics/math.[h/cpp]`: Wrappers for math functions that conform with those present in the `std` namespace. 
- [ ] `numerics/number_visitor.[h/cpp]`: A dictionary-type visitor class, including CSE support. To be used in the context of batch optimisation.
- [ ] `numerics/optimizer.[h/cpp]`:  Implementation of a (unified) batch optimiser. This takes in a collection of symbolic functions and their input arguments, and applies a selected method of optimisation to them to speed up evaluation. For numerical codes, this class can be extended to give support for tensor-based operations, for example.
- Tests in 'tests/numerics' covering features of all of the above.

If you agree that these are features that are useful to have in the core code of SymEngine, please let me know and I'll start upstreaming them. Please also let me know if you disagree with the proposed directory/namespace structure, as I would want to correct this as early on in the process as possible. Thanks for considering my proposal!

@luca-heltai ping

**Comment by certik** [comment]
@jppelteret I think this is a great idea. @isuruf has the best understanding of the low level details and it seems you have been in touch with him.

In general regarding the core of SymEngine, we didn't use operator overloading, and rather opted for explicit `add`/`mul`/... functions, because it was easier to deliver performance, const correctness etc. For expressions, we have a wrapper that implements the operator overloading:

https://github.com/symengine/symengine/blob/4cc711a935277c0fa93cb97f4b8280a497e08e89/symengine/expression.h

It might or might not have a performance overhead, we would have to carefully benchmark it, and I don't want to deal with this extra layer of complexity in the core. However, from the user API perspective, operator overloading is of course much simpler, and it makes sense to use it.

**Comment by jppelteret** [comment]
> For expressions, we have a wrapper that implements the operator overloading:

Ah, thank you for this reminder. @isuruf mentioned the `Expression` class at some point but I'd forgotten about it. I think that this and the `SymEngineNumber` that I've implemented (prior to my knowledge of `Expression`) are similar, but my current implementation is perhaps a bit more feature rich (for example, it allows the creation of function symbols, differentiation, and casting). Maybe we could combine the two implementations...

> In general regarding the core of SymEngine, we didn't use operator overloading, and rather opted for explicit `add`/`mul`/... functions, because it was easier to deliver performance, const correctness etc.

Since I now have a vague familiarity with how you've exploited this paradigm, I fully understand why you chose to go this route. My extensions are almost exclusively motivated because, in the context of most C++ tensor libraries, I have found they expect "numbers" to have the standard operators and math functions defined. Our library is no different in this regard.

Irrespective of the overhead, for our purposes the resulting symbolic expressions are so large that (assuming that the symbolic expressions are computed _a priori_ or once during simulation) their evaluation becomes the primary bottleneck. This is where the `BatchOptimizer` becomes really useful, and what I tend to want to exploit most readily.

If it would be of any use to you, I could push a branch that has the complete implementation in place so that you could a better sense of what I'd like to contribute?

**Comment by certik** [comment]
@jppelteret Yes, if you could push the branch, then we can see how to best merge things with symengine.

Yes, for your use case, the possible small penalty of the wrapper seems justifiable.

**Comment by jppelteret** [comment]
Ok I will do so. For the purposes of review I would be happy to split the implementation into smaller PR's.
